### DISCUSSION TITLE
Double tap p to dollar working but not ideal .. currently requires delete or backspace.

### DISCUSSION THREAD

**Discussion by momomo** [pqrs-org/Karabiner-Elements#4047] [General]
I have  the following below, with an attempt to output p for normal typing, and $ for double tap. 

This works, but currently requires me to delete the p that is being as well as the dollar output. 

Is there another way? Tested a bunch of options. 

If you propose something, please make sure to test it first as one might think something works, but in reality Karabiner does not behave as one expect it to do. 

```

         ///////////////////////////////////////////////////////////////////////////
        // Double p to dollar. 
        // However was not able to pull it off without delete_or_backspace.   
        ///////////////////////////////////////////////////////////////////////////   
        {
                "type": "basic",
                "conditions": [
                        { "type": "variable_if", "name": "mmm.p.dollar", "value": 1 }
                ],
                
                "from": { "key_code": "p" },
                
                "to": [ 
                        { "key_code": "delete_or_backspace", "modifiers": [ "left_shift" ] }, 
                        { "key_code": "4", "modifiers": [ "left_shift" ] } 
                ],
                
                "to_after_key_up": [
                        { "set_variable": { "name": "mmm.p.dollar", "value": 0 } }
                ]
        },                                                                 
        {
                "type": "basic",
                "from": { "key_code": "p" },
                
                "to": [  
                        { "set_variable": { "name": "mmm.p.dollar", "value": 1 } },
                        { "key_code": "p" }
                        
                ],
                
                "to_delayed_action": {
                        "to_if_invoked": [
                                { "set_variable": { "name": "mmm.p.dollar", "value": 0 } }
                        ],
                        "to_if_canceled": [
                                { "set_variable": { "name": "mmm.p.dollar", "value": 0 } }
                        ]
                },
                
                "parameters": { "basic.to_delayed_action_delay_milliseconds": 150 }
        }
```

**Comment by tekezo** [comment]
Thanks for the feedback. This requires a feature introduced in the beta version.
I’ve added an example, so please try replicating it.

https://karabiner-elements.pqrs.org/docs/json/typical-complex-modifications-examples/#change-double-press-of-q-to-escape

## How to upgrade to the latest beta version:
![Check for beta updates](https://karabiner-elements.pqrs.org/docs/manual/operation/check-for-updates/images/karabiner-elements-check-for-updates-beta@2x.png)

**Comment by momomo** [comment]
@tekezo 

That update overwrote my entire karabiner file! 

{
    "global": {
        "check_for_updates_on_startup": false,
        "show_profile_name_in_menu_bar": true
    },
    "profiles": [
        {
            "name": "Default profile",
            "selected": true,
            "virtual_hid_keyboard": { "keyboard_type_v2": "ansi" }
        }
    ]
}

I should be able to restore it but not great!

**Comment by momomo** [comment]
It does even when I replace it and restart the application!

**Comment by momomo** [comment]
I think that when it detects an error on startup on teh karabiner file it will default to cleaning it away. 

This is an issue that only works for first time installers or launchers. 

If there is already a working karabiner version, karabiner will not switch nor replace. Just log the error. 

But if you are launching karabiner and there is an error, bye bye! 

Probably should be fixed!

**Comment by momomo** [comment]
This is working: 

```

        ///////////////////////////////////////////////////////////////////////////
        // Double p to dollar. 
        // However was not able to pull it off without delete_or_backspace.
        ///////////////////////////////////////////////////////////////////////////   
        {
                "type": "basic",
                "conditions": [
                        { "type": "variable_if", "name": "mmm.p.dollar", "value": true }
                ],
                
                "from": {
                        "key_code": "p"
                },
                
                "to": [
                        { "set_variable": { "name": "mmm.p.dollar", "value": false } },
                        { "key_code": "4", "modifiers": [ "left_shift" ] }
                ]
                
        },
        
        {
                "type": "basic",
                "from": {
                        "key_code": "p"
                },
                
                "to": [{ "set_variable": { "name": "mmm.p.dollar", "value": true } }],
                
                "to_delayed_action": {
                        "to_if_invoked": [
                                {
                                        "key_code": "p",
                                        "conditions": [
                                                {
                                                        "type": "variable_if",
                                                        "name": "mmm.p.dollar",
                                                        "value": true
                                                }
                                        ]
                                },
                                { "set_variable": { "name": "mmm.p.dollar", "value": false } }
                        ],
                        "to_if_canceled": [
                                {
                                        "key_code": "p",
                                        "conditions": [
                                                {
                                                        "type": "variable_if",
                                                        "name": "mmm.p.dollar",
                                                        "value": true
                                                }
                                        ]
                                },
                                { "set_variable": { "name": "mmm.p.dollar", "value": false } }
                        ]
                },
                
                "parameters": { "basic.to_delayed_action_delay_milliseconds": 175 }
        }
        
        ```
        
        Note that the optional:['any']  in the example does no good and leads to weird behaviour ... for this case i don't see why any kind of combo should be allowed.

**Comment by momomo** [comment]
@tekezo How come karabiner does not su$port ( first issue with the idea of double p encountered!) the output of say swedish characters? 

Is there soem technical limitation as to why that doesn't work? 

Can one not provide utf or unicode chars? I see in the docs output of japanese characters work, but I couldn't get that to work either!

**Reply by tekezo** [reply]
It's simply a matter of differing responsibilities.

The role of a keyboard device is to pass scancodes to the OS, and interpreting those scancodes as specific characters is the responsibility of macOS's Input Source.
Since Karabiner-Elements functions as a virtual keyboard device, it only sends scancodes and doesn't inherently carry character information. This means that even Japanese characters cannot be directly output.

Therefore, character output must be handled on the input source side, or scancodes that the input source can interpret to produce the desired character must be sent from Karabiner-Elements. For example, sending option+t to output †.

**Reply by momomo** [reply]
But why don't you allow us to send any scancode by us to be interpreted OS, like relevant scan code out for Swedish, or Japanese characters then? 

One can then figure out what they are and type them in from a de-serializable string. 

I can get a keyboard that comes with ÅÄÖ ... so karabiner should be able to be such a virtual keyboard, no? 

Although I understand your point, but it would be nice if karabiner was the ultimate force. Its much more rigid to use ukulele where karabiner is a good tool to iterate with. 

The thing is that currently also karabiner limits the amount of target keys we can use to map things to ... and we have four mods, shift, control, option, command to play with ... any combo of these might lead to conflicts when setting up a super or severalother mod keys for different layers.

**Reply by tekezo** [reply]
Character processing isn't the responsibility of the keyboard, which means there are no scancodes directly corresponding to specific characters.

The most flexible way to input arbitrary characters would be to use Unicode Hex Input as Input Source and enter Unicode values directly. That said, the usability of this approach isn't exactly ideal.

**Comment by momomo** [comment]
I changed it to slash ... however I noticed that on double tap and hold ... the output leads to many dollars ... it should be on release only  and only if not previously held down.

**Comment by momomo** [comment]
Latest working version: 

```
        ///////////////////////////////////////////////////////////////////////////
        // Double quote to dollar
        ///////////////////////////////////////////////////////////////////////////
           
        {
                "conditions": [ { "type": "variable_if", "name": "mmm.p.dollar", "value": true } ],
                
                "from": { "key_code": "quote" },
                
                "to": [{ "set_variable": { "name": "mmm.p.dollar", "value": false } }, { "key_code": "4", "modifiers": [ "left_shift" ] }], "type": "basic"
        },
        
        {
                "from": { "key_code": "quote" },
                
                "to": [{ "set_variable": { "name": "mmm.p.dollar", "value": true } }],
                
                "to_delayed_action": {
                        "to_if_invoked": [
                                { "key_code": "quote", "conditions": [{"type": "variable_if", "name": "mmm.p.dollar", "value": true}]},
                                { "set_variable": { "name": "mmm.p.dollar", "value": false } }
                        ],
                        "to_if_canceled": [
                                { "key_code": "quote", "conditions": [{"type": "variable_if", "name": "mmm.p.dollar", "value": true}]},
                                { "set_variable": { "name": "mmm.p.dollar", "value": false } }
                        ]
                },
                
                "parameters": { "basic.to_delayed_action_delay_milliseconds": 150 }, "type": "basic"
        }

```
